codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -256,19 +255,36 @@ void isActiveCountChangeForTree(int increment) {
             activeCountForTree += increment;
             if (!stream.isRoot()) {
                 State pState = state(stream.parent());
+                assert activeCountForTree != increment ||
+                       priorityQueueIndex == INDEX_NOT_IN_QUEUE ||
+                       pState.queue.contains(this) :
+                     "State[" + stream.id() + "].activeCountForTree changed from 0 to " + increment + " is in a queue" +
+                     ", but not in parent[ " + pState.stream.id() + "]'s queue";
                 if (activeCountForTree == 0) {
                     pState.remove(this);
-                } else if (activeCountForTree - increment == 0) { // if frame count was 0 but is now not, then queue.
+                } else if (activeCountForTree == increment && !isInProcess()) {
+                    // If frame count was 0 but is now not, and this node is not already in a queue (assumed to be
+                    // pState's queue) then enqueue it. If this State object is being processed the pseudoTime for this
+                    // node should not be adjusted, and the node will be added back to the queue/tree structure after it
+                    // is done being processed. This may happen if the activeCountForTree == 0 (a node which can't
```

Comment:
If `isDistributing()` (new name) is true that means the `State` object has been removed from its parent's queue to be processed, and will be added back (if necessary) after it has been processed. So I don't think it is a problem if this method is called while the `State` object is being processed, because it already won't be in the queue so `queue.remove(..)` should be a noop.

It is possible for a element to be remove in a transient manner (during a tree shift), but it will be added back after the tree shift operation is complete if it has children which are "active" ... this is what happens during the new unit test.